src/api-service/__app__/timer_retention/__init__.py

```diff
@@ -0,0 +1,97 @@
+import datetime
+import logging
+
+import azure.functions as func
+from onefuzztypes.enums import JobState, TaskState
+
+from ..onefuzzlib.events import get_events
+from ..onefuzzlib.jobs import Job
+from ..onefuzzlib.notifications.main import Notification
+from ..onefuzzlib.repro import Repro
+from ..onefuzzlib.tasks.main import Task
+
+RETENTION_POLICY = datetime.timedelta(minutes=(5))
+
+
+def main(mytimer1: func.TimerRequest, dashboard: func.Out[str]) -> None:  # noqa: F841
+
+    now = datetime.datetime.now(tz=datetime.timezone.utc)
+
+    time_retained = now - RETENTION_POLICY
+
+    time_filter = f"Timestamp lt datetime'{time_retained.isoformat()}'"
+
+    # You have to do notification before task,
+    # because editing the upn for tasks will change the timestamp
+    for notification in Notification.search(raw_unchecked_filter=time_filter):
+        logging.info("Retention Timer Notification Search")
+        logging.info(
+            "Found notification %s older than 18 months. Checking related tasks.",
+            notification.notification_id,
+        )
+        container = notification.container
+        timestamp_list = []
+        for task in Task.search():
+            container_str = str(task.config.containers)
+            if container in container_str:
+                # Need to make sure there isn't a task still using the container.
+                if task.state == TaskState.stopped:
+                    timestamp_list.append(task.timestamp)
+                else:
+                    timestamp_list = []
+                    break
+
+        if len(timestamp_list) != 0:
+            youngest = max(
+                dt
+                for dt in timestamp_list
+                if isinstance(dt, datetime.datetime) and dt < now
+            )
+            if youngest < time_retained:
+                logging.info(
+                    "All related tasks are older than 18 months."
+                    + " Deleting Notification %s.",
+                    notification.notification_id,
+                )
+                notification.delete()
+
+    for job in Job.search(

```

Comment:
For a system that is always in use, this will progressively get slower.  Perhaps the time filter can be "older than 180 days, newer than 200 days" as to not recheck every job older than 180 days each time this is run?

Reply:
This makes sense!